server/src/main/java/io/druid/client/CoordinatorServerView.java

```diff
@@ -0,0 +1,213 @@
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.VersionedIntervalTimeline;
+import io.druid.timeline.partition.PartitionChunk;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * ServerView of coordinator for the state of segments being loaded in the cluster.
+ */
+public class CoordinatorServerView implements InventoryView
+{
+  private static final Logger log = new Logger(CoordinatorServerView.class);
+
+  private final Object lock = new Object();
+
+  private final Map<String, Set<DruidServerMetadata>> servers;
+  private final Map<String, VersionedIntervalTimeline<String, SegmentLoadInfo>> timelines;
+
+  private final ServerInventoryView baseView;
+
+  private volatile boolean initialized = false;
+
+  @Inject
+  public CoordinatorServerView(
+      ServerInventoryView baseView
+  )
+  {
+    this.baseView = baseView;
+    this.servers = Maps.newHashMap();
+    this.timelines = Maps.newHashMap();
+
+    ExecutorService exec = Execs.singleThreaded("CoordinatorServerView-%s");
+    baseView.registerSegmentCallback(
+        exec,
+        new ServerView.SegmentCallback()
+        {
+          @Override
+          public ServerView.CallbackAction segmentAdded(DruidServerMetadata server, DataSegment segment)
+          {
+            serverAddedSegment(server, segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentRemoved(final DruidServerMetadata server, DataSegment segment)
+          {
+            serverRemovedSegment(server, segment);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+
+          @Override
+          public ServerView.CallbackAction segmentViewInitialized()
+          {
+            initialized = true;
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+
+    baseView.registerServerCallback(
+        exec,
+        new ServerView.ServerCallback()
+        {
+          @Override
+          public ServerView.CallbackAction serverRemoved(DruidServer server)
+          {
+            removeServer(server);
+            return ServerView.CallbackAction.CONTINUE;
+          }
+        }
+    );
+  }
+
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  public void clear()
+  {
+    synchronized (lock) {
+      timelines.clear();
+      servers.clear();
+    }
+  }
+
+  private void removeServer(DruidServer server)
+  {
+    for (DataSegment segment : server.getSegments().values()) {
+      serverRemovedSegment(server.getMetadata(), segment);
+    }
+  }
+
+  private void serverAddedSegment(final DruidServerMetadata server, final DataSegment segment)
+  {
+    String segmentId = segment.getIdentifier();
+    synchronized (lock) {
+      log.debug("Adding segment[%s] for server[%s]", segment, server);
+
+      Set<DruidServerMetadata> servers = this.servers.get(segmentId);
```

Comment:
Can you please rename this, it's confusing to have servers here and also servers as a class member.